ghcide/src/Development/IDE/GHC/ExactPrint.hs

```diff
@@ -271,6 +346,44 @@ graftDecls dst decs0 = Graft $ \dflags a -> do
             | otherwise = DL.singleton (L src e) <> go rest
     modifyDeclsT (pure . DL.toList . go) a
 
+graftSmallestDecls ::
+    forall a.
+    (HasDecls a) =>
+    SrcSpan ->
+    [LHsDecl GhcPs] ->
+    Graft (Either String) a
+graftSmallestDecls dst decs0 = Graft $ \dflags a -> do
+    decs <- forM decs0 $ \decl -> do
+        (anns, decl') <- annotateDecl dflags decl
+        modifyAnnsT $ mappend anns
+        pure decl'
+    let go [] = DL.empty
+        go (L src e : rest)
+            | dst `isSubspanOf` src = DL.fromList decs <> DL.fromList rest
+            | otherwise = DL.singleton (L src e) <> go rest
+    modifyDeclsT (pure . DL.toList . go) a
```

Comment:
May I ask why do you use the difference list here? It seems B
both `DL.singleton x <> l` and `DL.fromList x <> DL.fromList y` are not more efficient than their list counterparts if we apply `DL.toList` immediately after...
I mean, this is basically
```
a : b : c : ... : ((x ++) . (y ++) $ [])
```
which is not better than
```
a : b : c : ... : (x ++ y)
```
right?

Reply:
Maybe it's copied from line 344-347? (I also don't see why that part uses `DList`)

Reply:
I just copied and pasted it from the other definition!

Reply:
@konn maybe you can chime in?

Reply:
Closing this, but feel free to file a bug and tag me in it if you feel strongly; I'm happy to clean up this module in a subsequent PR.